src/browser/ChromeListener.cpp

```diff
@@ -0,0 +1,614 @@
+        m_sd.close();
+    }
+
+    if (!m_io_service.stopped())
+        m_io_service.stop();
+
+    m_fut.waitForFinished();
+    m_running = false;
+}
+
+void ChromeListener::readDatagrams()
+{
+    QByteArray dgram;
+
+    while (m_udpSocket.hasPendingDatagrams()) {
+        dgram.resize(m_udpSocket.pendingDatagramSize());
+        m_udpSocket.readDatagram(dgram.data(), dgram.size(), &m_peerAddr, &m_peerPort);
+    }
+
+    QJsonParseError err;
+    QJsonDocument doc(QJsonDocument::fromJson(dgram, &err));
+    if (doc.isObject()) {
+        QJsonObject json = doc.object();
+        QString val = json.value("action").toString();
+        if (!val.isEmpty()) {
+            // Allow public keys to be changed without database being opened
+            if (val != "change-public-keys" && !m_service.isDatabaseOpened()) {
+                if (!m_service.openDatabase()) {
+                    sendErrorReply(val, ERROR_KEEPASS_DATABASE_NOT_OPENED);
+                }
+            }
+            else {
+                handleAction(json);
+            }
+        }
+    }
+}
+
+void ChromeListener::readHeader(boost::asio::posix::stream_descriptor& sd)
+{
+    char buf[4] = {};
+    async_read(sd, buffer(buf,sizeof(buf)), transfer_at_least(1), [&](error_code ec, size_t br) {
+        if (!ec && br >= 1) {
+            uint len = 0;
+            for (int i = 0; i < 4; i++) {
+                uint rc = buf[i];
+                len = len | (rc << i*8);
+            }
+            readBody(sd, len);
+        }
+    });
+}
+
+void ChromeListener::readBody(boost::asio::posix::stream_descriptor& sd, const size_t len)
+{
+    char buf[MESSAGE_LENGTH] = {};
+    async_read(sd, buffer(buf, len), transfer_at_least(1), [&](error_code ec, size_t br) {
+        if (!ec && br > 0) {
+            QByteArray arr(buf, br);
+            QJsonParseError err;
+            QJsonDocument doc(QJsonDocument::fromJson(arr, &err));
+            if (doc.isObject()) {
+                QJsonObject json = doc.object();
+                QString val = json.value("action").toString();
+                if (!val.isEmpty()) {
+                    // Allow public keys to be changed without database being opened
+                    if (val != "change-public-keys" && !m_service.isDatabaseOpened()) {
+                        if (!m_service.openDatabase()) {
+                            sendErrorReply(val, ERROR_KEEPASS_DATABASE_NOT_OPENED);
+                        }
+                    }
+                    else {
+                        handleAction(json);
+                    }
+                }
+            }
+            readHeader(sd);
+        }
+    });
+}
+
+void ChromeListener::readLine()
+{
+    // Read the message header
+    readHeader(m_sd);
+    m_io_service.run();
+}
+
+void ChromeListener::handleAction(const QJsonObject &json)
+{
+    QString action = json.value("action").toString();
+    if (!action.isEmpty()) {
+        if (action == "get-databasehash")
```

Comment:
You could use qHash(). But on the other hand not such a big deal. You can leave it as is.